### PR TITLE
Add shap support for linear models

### PR DESCRIPTION
**Pull request by anilkumarpanda**
- Add support for SHAPley value calculation for linear models. The current setup does not support models like `SVC(kernel='rbf') or KNeighborsClassifier or is very slow.  I will create a separate issue for adding that.
- Upgraded shap version to 0.38.1

### CODE REVIEW

File: probatus/utils/shap_helpers.py [hunk 1]
```diff
@@ -62,9 +60,18 @@ def shap_calc(model, X, approximate=False, return_explainer=False, verbose=0, **
...
+        # Create the background data,required for non tree based models.
+        # A single datapoint can passed as mask (https://github.com/slundberg/shap/issues/955#issuecomment-569837201)
+        
+        if X.shape[1]< sample_size :
+            sample_size = int(np.ceil(X.shape[1]*0.2))
+        else :
+           pass
+        mask = shap.utils.sample(X,sample_size)
+
+        explainer = shap.Explainer(model,masker=mask,**shap_kwargs)
```
Comment (sbjelogr) [inline probatus/utils/shap_helpers.py:72]: Isn't this step a bit dangerous?
If we pass a tree based model, this would still use the Explainer (instead of the Tree Explainer), which is not something you really want, right?

Comment (Matgrb) [reply probatus/utils/shap_helpers.py:72]: I think the new shap basically selects the explainer to be used based on the provided model.

If you pass tree model it will use tree explainer within Explainer

Comment (anilkumarpanda) [reply probatus/utils/shap_helpers.py:72]: With the latest version of SHAP ,[Explainer](https://shap.readthedocs.io/en/latest/generated/shap.Explainer.html#shap.Explainer.__init__) by default attempts to make the best choice given the passed model. So this spares identifying the model type in the code.
However,  having said that I think models like SVC(kernel) and KNN etc needs to handled separately.

File: docs/tutorials/nb_shap_feature_elimination.ipynb [hunk 2]
```diff
@@ -8,7 +8,7 @@
     "\n",
     "[![open in colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/ing-bank/probatus/blob/master/docs/tutorials/nb_shap_feature_elimination.ipynb)\n",
     "\n",
-    "Recursive Feature Elimination allows to efficiently reduce the number of features in your dataset, without losing the predictive power of the model. `probatus` implements the following feature elimination routine for **tree-based models**:\n",
+    "Recursive Feature Elimination allows to efficiently reduce the number of features in your dataset, without losing the predictive power of the model. `probatus` implements the following feature elimination routine for **tree-based,linear models**:\n",
```
Comment (Matgrb) [inline docs/tutorials/nb_shap_feature_elimination.ipynb:11]: Isn't it for any provided model?

Good pick are often tree-based or linear, but can be also other.

File: docs/tutorials/nb_shap_feature_elimination.ipynb [hunk 3]
```diff
@@ -22,15 +22,15 @@
     " \n",
     "The main advantages of using this routine are:\n",
     "\n",
-    "- It uses a tree-based model to detect the complex relations between features and the target.\n",
+    "- It uses a tree-based,linear model to detect the complex relations between features and the target.\n",
```
Comment (Matgrb) [inline docs/tutorials/nb_shap_feature_elimination.ipynb:25]: Same comment as before

File: docs/tutorials/nb_shap_feature_elimination.ipynb [hunk 4]
```diff
@@ -22,15 +22,15 @@
...
     "- You can also provide a list of features, that should not be eliminated. E.g incase of prior knowledge.\n",
     "\n",
     "The disadvantages are:\n",
     "\n",
     "- Removing lowest [SHAP](https://shap.readthedocs.io/en/latest/) importance feature does not always translate to choosing the feature with lowest impact on model's performance. Shap importance illustrates how strongly a given feature affects the output of the model, while disregarding correctness of this prediction.\n",
-    "- Currently, the functionality only supports tree-based binary classifiers, in the future the scope might be extended.\n",
+    "- Currently, the functionality only supports tree-based,linear binary classifiers, in the future the scope might be extended.\n",
```
Comment (Matgrb) [inline docs/tutorials/nb_shap_feature_elimination.ipynb:33]: wide range of models, however, the best picks are tree-based or linear.

I would rephrase it.

File: tests/feature_elimination/test_feature_elimination.py [hunk 5]
```diff
@@ -73,6 +76,53 @@ def test_shap_rfe(X, y, capsys):
...
+
+    # Ensure that number of warnings was 0
+    assert len(record) == 0
+    # Check if there is any prints
+    out, _ = capsys.readouterr()
+    assert len(out) == 0
+
+def test_shap_rfe_svm(X, y, capsys):
+    
+    clf = SVC(C=1,kernel='linear',probability=True)
```
Comment (Matgrb) [inline tests/feature_elimination/test_feature_elimination.py:104]: Can you add a test with non-linear model? like KNN or SVC with rbf kernel

File: tests/feature_elimination/test_feature_elimination.py [hunk 6]
```diff
@@ -153,9 +203,9 @@ def test_get_feature_shap_values_per_fold(X, y):
 
 @pytest.mark.skipif(os.environ.get("SKIP_LIGHTGBM") == 'true', reason="LightGBM tests disabled")
 def test_complex_dataset(complex_data, complex_lightgbm):
+    
     X, y = complex_data
 
-
     param_grid = {
         'n_estimators': [5, 7, 10],
         'num_leaves': [3, 5, 7, 10],
```
Comment (Matgrb) [inline tests/feature_elimination/test_feature_elimination.py:211]: I would also add tests with Linear and non-linear models to the Shap resemblance model and Shap interpreter.

**Review by Matgrb — CHANGES_REQUESTED** [review]
Overall looks good, I suggested a couple of small adjustments, let me know what do you think about them

File: docs/tutorials/nb_shap_feature_elimination.ipynb (hunk 2 above)
Comment (anilkumarpanda) [reply docs/tutorials/nb_shap_feature_elimination.ipynb:11]: Theoretical yes, however the there is an issue in SHAP when you pass model like KNN or SVM(kernel='rbf') it throws an error. There is an open [issue about it](https://github.com/slundberg/shap/issues/1526) . We can mention now what models work and later update it.

Comment (Matgrb) [reply docs/tutorials/nb_shap_feature_elimination.ipynb:11]: Okay, then lets add a todo to include it once SHAP updates the version again

Comment (anilkumarpanda) [reply docs/tutorials/nb_shap_feature_elimination.ipynb:11]: Will create an issue about it.

File: tests/feature_elimination/test_feature_elimination.py (hunk 6 above)
Comment (anilkumarpanda) [reply tests/feature_elimination/test_feature_elimination.py:211]: Added tests to the interpret module and resemblance modules.

**Comment by anilkumarpanda** [comment]
Resolves #61 as well.

File: docs/tutorials/nb_shap_feature_elimination.ipynb (hunk 2 above)
Comment (Matgrb) [inline docs/tutorials/nb_shap_feature_elimination.ipynb:11]: Micro fix for each notebook:
tree-based **and** linear models

**Review by Matgrb — APPROVED** [review]
Great work! Thanks for implementing this!

Left a small comment,  but apart from that ready to merge.
